nano_graphrag/_spliter.py

```diff
@@ -0,0 +1,121 @@
+
+    def _split_tokens_with_separators(self, tokens: List[int]) -> List[List[int]]:
+        splits = []
+        current_split = []
+        i = 0
+        while i < len(tokens):
+            separator_found = False
+            for separator in self._separators:
+                if tokens[i:i+len(separator)] == separator:
+                    if current_split:
+                        if self._keep_separator == "end":
+                            current_split.extend(separator)
+                            splits.append(current_split)
+                            current_split = []
+                        elif self._keep_separator == "start":
+                            splits.append(current_split)
+                            current_split = separator[:]
+                        else:
+                            splits.append(current_split)
+                            current_split = []
+                    elif self._keep_separator:
+                        current_split.extend(separator)
+                    i += len(separator)
+                    separator_found = True
+                    break
+            if not separator_found:
+                current_split.append(tokens[i])
+                i += 1
+        if current_split:
+            splits.append(current_split)
+        return [s for s in splits if s]
+
+    def _merge_splits(self, splits: List[List[int]]) -> List[List[int]]:
+        merged_splits = []
+        current_split = []
+        current_length = 0
+        separator = [] if self._keep_separator is False else self._separators[-1]
+
+        for split in splits:
+            if self._length_function(current_split) + self._length_function(split) <= self._chunk_size:
+                if current_split and separator:
+                    current_split.extend(separator)
+                current_split.extend(split)
+            else:
+                if current_split:
+                    merged_splits.append(current_split)
+                current_split = split
+            if self._length_function(current_split) >= self._chunk_size:
+                merged_splits.append(current_split)
+                current_split = []
+        if current_split:
+            merged_splits.append(current_split)
+
+        if self._chunk_overlap > 0:
+            return self._enforce_overlap(merged_splits)
+        return merged_splits
+
+    def _enforce_overlap(self, chunks: List[List[int]]) -> List[List[int]]:
+        new_chunks = []
+        for i, chunk in enumerate(chunks):
+            if i == 0:
+                new_chunks.append(chunk)
+            else:
+                overlap_tokens = chunks[i-1][-self._chunk_overlap:]
+                new_chunk = overlap_tokens + chunk
+                if self._length_function(new_chunk) > self._chunk_size:
+                    new_chunk = new_chunk[-self._chunk_size:]
+                new_chunks.append(new_chunk)
+        return new_chunks
+
+# EXAMPLE USAGE
+if __name__ == "__main__":
```

Comment:
May be move this section to `tests`? Make sure the splitter is tested

Reply:
OK. Let me try this.

Reply:
done

Reply:
oh god wait

Reply:
done